gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/JWKSResource.java

```diff
@@ -0,0 +1,108 @@
+import com.nimbusds.jose.jwk.RSAKey;
+import org.apache.knox.gateway.config.GatewayConfig;
+import org.apache.knox.gateway.services.GatewayServices;
+import org.apache.knox.gateway.services.ServiceType;
+import org.apache.knox.gateway.services.security.KeystoreService;
+import org.apache.knox.gateway.services.security.KeystoreServiceException;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Singleton;
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import java.security.KeyStoreException;
+import java.security.interfaces.RSAPublicKey;
+
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+@Singleton
+@Path(JWKSResource.RESOURCE_PATH)
+public class JWKSResource {
+
+  static final String RESOURCE_PATH = "knoxtoken/api/v1/.well-known";
+  static final String JWKS_PATH = "/jwks.json";
+  @Context
+  HttpServletRequest request;
+  @Context
+  ServletContext context;
+  private KeystoreService keystoreService;
+
+  @PostConstruct
+  public void init() {
+    final GatewayServices services = (GatewayServices) context
+        .getAttribute(GatewayServices.GATEWAY_SERVICES_ATTRIBUTE);
+    keystoreService = services.getService(ServiceType.KEYSTORE_SERVICE);
+  }
+
+  @GET
+  @Path(JWKS_PATH)
+  @Produces({ APPLICATION_JSON })
+  public Response getJwksResponse() {
+    return getJwks(null);
+  }
+
+  private Response getJwks(final String keystore) {
+    JWKSet jwks;
+    try {
+      final RSAPublicKey rsa = getPublicKey(keystore);
+      final RSAKey.Builder builder = new RSAKey.Builder(rsa)
+          .keyUse(KeyUse.SIGNATURE)
+          .algorithm(new JWSAlgorithm(rsa.getAlgorithm()))
+          .keyIDFromThumbprint();
+
+      jwks = new JWKSet(builder.build());
+
+    } catch (KeyStoreException | JOSEException e) {
+      return Response.status(500)
+          .entity("{\n  \"error\": \"" + e.toString() + "\"\n}\n").build();
+    } catch (KeystoreServiceException e) {
+      return Response.status(500).entity(
+          "{\n  \"error\": \"" + "keystore " + keystore + " could not be found."
+              + "\"\n}\n").build();
+    }
+    return Response.ok()
+        .entity(jwks.toJSONObject().toString()).build();
+  }
+
+  protected RSAPublicKey getPublicKey(final String keystore)
+      throws KeystoreServiceException, KeyStoreException {
+    return (RSAPublicKey) keystoreService.getSigningKeystore(keystore)
+        .getCertificate(getSigningKeyAlias()).getPublicKey();
+  }
+
+  private String getSigningKeyAlias() {
+    final GatewayConfig config = (GatewayConfig) request.getServletContext()
+        .getAttribute(GatewayConfig.GATEWAY_CONFIG_ATTRIBUTE);
+    final String alias = config.getSigningKeyAlias();
+    return (alias == null) ? GatewayConfig.DEFAULT_SIGNING_KEY_ALIAS : alias;
```

Comment:
should getSigningKeyAlias be changed to return the default if it isn't explicitly set? Is that all this is doing?

Reply:
I do not fully follow the first question, do you mean change the `config.getSigningKeyAlias()` method?
But this method is just returning default signing key alias in case none is found.